### PR TITLE
Add gateway for exposing Intune Enrollment IDs

### PR DESCRIPTION
**Pull request by midella**
Add gateway for exposing Intune Enrollment IDs

### CODE REVIEW

File: ADAL/tests/unit/ios/ADEnrollmentGatewayTests.m [hunk 1]
```diff
@@ -0,0 +1,109 @@
...
+                                         "\"unique_account_id\" : \"1e4dd613-dave-4527-b50a-97aca38b57ba\",\n"
+                                         "\"user_id\" : \"dave@contoso.com\",\n"
+                                         "\"enrollment_id\" : \"64d0557f-dave-4193-b630-8491ffd3b180\"\n"
+                                     "},\n"
+                                    "]\n"
+            "}"];
+
+    // mock out enrollment data
+
+    Class ADEnrollmentGatewayClass = [ADEnrollmentGateway class];
```
Comment (oldalton) [inline ADAL/tests/unit/ios/ADEnrollmentGatewayTests.m]: I'd prefer to use e.g. category or have a private method to set test contents instead of using swizzling here.

File: ADAL/tests/unit/ios/ADEnrollmentGatewayTests.m [hunk 2]
```diff
@@ -0,0 +1,109 @@
...
+    // restore mocked method
+
+    Class ADEnrollmentGatewayClass = [ADEnrollmentGateway class];
+    SEL allEnrollmentIDsSelector = @selector(allEnrollmentIds);
+    Method allEnrollmentIDsMethod = class_getClassMethod(ADEnrollmentGatewayClass, allEnrollmentIDsSelector);
+
+    method_setImplementation(allEnrollmentIDsMethod, self.originalAllEnrollmentIds);
+}
+
+- (void) testenrollmentIdForUserId
```
Comment (oldalton) [inline ADAL/tests/unit/ios/ADEnrollmentGatewayTests.m]: please also add test cases for when enrollment ids dictionary is empty or corrupted

File: ADAL/src/ADEnrollmentGateway.m [hunk 3]
```diff
@@ -0,0 +1,161 @@
...
+    if (!enrollIdJSON)
+        return nil;
+
+    NSError* error = nil;
+    id enrollIds = [NSJSONSerialization JSONObjectWithData:[enrollIdJSON dataUsingEncoding:NSUTF8StringEncoding] options:kNilOptions error:&error];
+
+    if (error)
+        return nil;
+
+    enrollIds = enrollIds[enrollmentIdArray];
```
Comment (oldalton) [inline ADAL/src/ADEnrollmentGateway.m]: please also validate that enrollIds is valid and that enrollIds is actually a dictionary, because any app can actually write to user defaults and can corrupt them.

File: ADAL/src/ADEnrollmentGateway.m [hunk 4]
```diff
@@ -0,0 +1,161 @@
...
+    if (error)
+        return nil;
+
+    return resources[[ADEnrollmentGateway normalizeAuthority:authority]];
+}
+
+#if AD_BROKER
++ (void)setIntuneMamResourceWithJsonBlob:(NSString *)resources
+{
+    ADIntuneResourceJSON = [resources copy];
```
Comment (oldalton) [inline ADAL/src/ADEnrollmentGateway.m]: What's about thread safety?

File: ADAL/src/ADEnrollmentGateway.m [hunk 5]
```diff
@@ -0,0 +1,161 @@
...
+NSString* const user_id = @"user_id";
+NSString* const enrollment_id = @"enrollment_id";
+
+// Keys for Intune Resource
+NSString* const ADIntuneAppProtectionResourceID = @"intune_mam_resource_V";
+NSString* const ADIntuneAppProtectionResourceIDVersion = @"1";
+#define AD_INTUNE_RESOURCE_ID_KEY [ADIntuneAppProtectionEnrollmentID stringByAppendingString:ADIntuneAppProtectionEnrollmentIDVersion]
+
+static NSString* ADIntuneEnrollmentIdJSON = nil;
+static NSString* ADIntuneResourceJSON = nil;
```
Comment (jasoncoolmax) [inline ADAL/src/ADEnrollmentGateway.m]: Could we make the static variable something like `s_intuneResourceJSON` to be consistant with rest of the project? :)

File: ADAL/src/ADEnrollmentGateway.m [hunk 6]
```diff
@@ -0,0 +1,161 @@
...
+{
+    NSURL* authorityURL = [NSURL URLWithString:authority];
+    NSNumber *port = authorityURL.port;
+
+    // This assumes we're using https, which is mandatory for all AAD communications.
+    if (port == nil || port.intValue == 443)
+    {
+        return authorityURL.host.lowercaseString;
+    }
+    return [NSString stringWithFormat:@"%@:%d", authorityURL.host.lowercaseString, port.intValue];
```
[the remark this answers is not in the document]
Comment (jasoncoolmax) [reply ADAL/src/ADEnrollmentGateway.m]: here :)
https://github.com/AzureAD/azure-activedirectory-library-for-objc/blob/release/2.6.4/ADAL/src/utils/NSURL%2BADExtensions.m#L83

File: ADAL/tests/unit/ios/ADEnrollmentGatewayTests.m [hunk 7]
```diff
@@ -0,0 +1,182 @@
...
+    XCTAssertNil([ADEnrollmentGateway enrollmentIdForUniqueAccountId:@"60406d5d-mike-41e1-aa70-e97501076a22" error:NULL]);
+
+}
+
+- (void) testEnrollmentIDErrorsReturned
+{
+    // dictionary of dictionaries
+    [ADEnrollmentGateway setEnrollmentIdsWithJsonBlob:[NSString stringWithFormat:@"{\"enrollment_ids\":{\"enrollment_ids\":{\"enrollment_ids\":\"enrollment_ids\"}}}"]];
+
+    // Valid JSON does not return error
```
Comment (midella) [inline ADAL/tests/unit/ios/ADEnrollmentGatewayTests.m]: In this case, I could also generate a new error object and populate it with a message stating that the JSON doesn't conform to the expected structure, but I didn't see any existing examples of this pattern, so I left it as log at WARN level, let me know if I should change this.

Comment (jasoncoolmax) [reply ADAL/tests/unit/ios/ADEnrollmentGatewayTests.m]: here is an example:
https://github.com/AzureAD/azure-activedirectory-library-for-objc/blob/release/2.6.4/ADAL/src/request/ADAuthenticationRequest%2BAcquireToken.m#L207

We could add an addition error code, say `AD_ERROR_ENROLLMENT_IDS_BAD_FORMAT` :)

File: ADAL/src/ADAuthenticationError.m [hunk 8]
```diff
@@ -333,6 +333,7 @@ + (NSString*)stringForADErrorCode:(ADErrorCode)code
             AD_ERROR_CODE_ENUM_CASE(AD_ERROR_TOKENBROKER_RESPONSE_NOT_RECEIVED);
             AD_ERROR_CODE_ENUM_CASE(AD_ERROR_TOKENBROKER_FAILED_TO_CREATE_KEY);
             AD_ERROR_CODE_ENUM_CASE(AD_ERROR_TOKENBROKER_DECRYPTION_FAILED);
+            AD_ERROR_CODE_ENUM_CASE(AD_ERROR_DEVELOPER_JSON_DATA_STRUCTURE_MALFORMED);
```
Comment (jasoncoolmax) [inline ADAL/src/ADAuthenticationError.m]: I think the JSON is not provided by developer, right? :)

Comment (midella) [reply ADAL/src/ADAuthenticationError.m]: It should be provided by the MAM SDK, so that's correct, I'm not sure which other bucket it fits into though.

Comment (midella) [reply ADAL/src/ADAuthenticationError.m]: Hmm, we could just use AD_ERROR_UNEXPECTED for this code, as we'd only hit this if the app unexpectedly modifies NSUserDefaults

File: ADAL/src/ADEnrollmentGateway.m [hunk 9]
```diff
@@ -0,0 +1,212 @@
...
+#define AD_INTUNE_ENROLLMENT_ID @"intune_app_protection_enrollment_id_V"
+#define AD_INTUNE_ENROLLMENT_ID_VERSION @"1"
+#define AD_INTUNE_ENROLLMENT_ID_KEY (AD_INTUNE_ENROLLMENT_ID AD_INTUNE_ENROLLMENT_ID_VERSION)
+
+// Keys for Intune Resource
+#define AD_INTUNE_RESOURCE_ID @"intune_mam_resource_V"
+#define AD_INTUNE_RESOURCE_ID_VERSION @"1"
+#define AD_INTUNE_RESOURCE_ID_KEY (AD_INTUNE_RESOURCE_ID AD_INTUNE_RESOURCE_ID_VERSION)
+
+NSString* const enrollmentIdArray = @"enrollment_ids";
```
Comment (jasoncoolmax) [inline ADAL/src/ADEnrollmentGateway.m]: shall we rename it `ENROLLMENT_ID_ARRAY` as the constants below?

File: ADAL/src/ADAuthenticationError.m (hunk 8 above)
Comment (jasoncoolmax) [reply ADAL/src/ADAuthenticationError.m]: Talked to Olga, she agreed that we can use AD_ERROR_UNEXPECTED :)

File: ADAL/src/ADEnrollmentGateway.m [hunk 10]
```diff
@@ -0,0 +1,227 @@
...
+#define AD_INTUNE_ENROLLMENT_ID @"intune_app_protection_enrollment_id_V"
+#define AD_INTUNE_ENROLLMENT_ID_VERSION @"1"
+#define AD_INTUNE_ENROLLMENT_ID_KEY (AD_INTUNE_ENROLLMENT_ID AD_INTUNE_ENROLLMENT_ID_VERSION)
+
+// Keys for Intune Resource
+#define AD_INTUNE_RESOURCE_ID @"intune_mam_resource_V"
+#define AD_INTUNE_RESOURCE_ID_VERSION @"1"
+#define AD_INTUNE_RESOURCE_ID_KEY (AD_INTUNE_RESOURCE_ID AD_INTUNE_RESOURCE_ID_VERSION)
+
+NSString * const ENROLLMENT_ID_ARRAY = @"enrollment_ids";
```
Comment (jasoncoolmax) [inline ADAL/src/ADEnrollmentGateway.m]: I talked to Olga last Friday, and she suggested that we could stick with the style `NSString *const`, as it is already used in our new code :)

File: ADAL/src/ADEnrollmentGateway.h [hunk 11]
```diff
@@ -0,0 +1,38 @@
...
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+@interface ADEnrollmentGateway : NSObject
+
++ (NSString *)enrollmentIdForUserId:(NSString *)userId error:(ADAuthenticationError*__autoreleasing *) error;
```
Comment (jasoncoolmax) [inline ADAL/src/ADEnrollmentGateway.h]: code style nit: 
`(ADAuthenticationError*__autoreleasing *) error` to 
`(ADAuthenticationError * __autoreleasing *)error`

File: ADAL/src/ADEnrollmentGateway.h [hunk 12]
```diff
@@ -0,0 +1,38 @@
...
++ (NSString *)enrollmentIdForUserObjectId:(NSString *)userObjectId tenantId:(NSString *)tenantId error:(ADAuthenticationError*__autoreleasing *) error;
```
Comment (jasoncoolmax) [inline ADAL/src/ADEnrollmentGateway.h]: code style nit: 
remove space between `)` and `error` :)

File: ADAL/src/ADEnrollmentGateway.h [hunk 13]
```diff
@@ -0,0 +1,38 @@
...
++ (NSString *)enrollmentIdForUniqueAccountId:(NSString *)uniqueAccountId error:(ADAuthenticationError*__autoreleasing *) error;
```
Comment (jasoncoolmax) [inline ADAL/src/ADEnrollmentGateway.h]: code style nit: 
remove space between `)` and `error` :)

File: ADAL/src/ADEnrollmentGateway.h [hunk 14]
```diff
@@ -0,0 +1,38 @@
...
++ (NSString *)allEnrollmentIds;
+
++ (NSString *)intuneMamResource:(NSURL *)authority error:(ADAuthenticationError*__autoreleasing *) error;
```
Comment (jasoncoolmax) [inline ADAL/src/ADEnrollmentGateway.h]: same code style nit as above

File: ADAL/src/ADEnrollmentGateway.m [hunk 15]
```diff
@@ -0,0 +1,227 @@
...
+NSString *const USER_ID = @"user_id";
+NSString *const ENROLLMENT_ID = @"enrollment_id";
+
+static NSString *s_intuneEnrollmentIdJSON = nil;
+static NSString *s_intuneResourceJSON = nil;
+
+
+@interface ADEnrollmentGateway()
+
++ (NSString *) getEnrollmentIDForIdentifier:(BOOL (^)(NSDictionary*)) idBlock error:(ADAuthenticationError *__autoreleasing *) error;
```
Comment (jasoncoolmax) [inline ADAL/src/ADEnrollmentGateway.m]: same code style nit as above

File: ADAL/src/ADEnrollmentGateway.m [hunk 16]
```diff
@@ -0,0 +1,227 @@
...
+
++ (NSString *)enrollmentIdForUserObjectId:(NSString *)userObjectId tenantId:(NSString *)tenantId error:(ADAuthenticationError *__autoreleasing *) error
+{
+    return [ADEnrollmentGateway getEnrollmentIDForIdentifier:^BOOL(NSDictionary *dic) {
+        return [[dic objectForKey:OID] isEqualToString:userObjectId] && [[dic objectForKey:TID] isEqualToString:tenantId];
+    }
+                                                       error:error];
+}
+
++ (NSString *)enrollmentIdForUniqueAccountId:(NSString *)uniqueAccountId error:(ADAuthenticationError *__autoreleasing *) error
```
Comment (jasoncoolmax) [inline ADAL/src/ADEnrollmentGateway.m]: same code style nit as above, also remove space after the right parenthesis :)

File: ADAL/src/ADEnrollmentGateway.m [hunk 17]
```diff
@@ -0,0 +1,227 @@
...
+{
+    return [ADEnrollmentGateway getEnrollmentIDForIdentifier:^BOOL(NSDictionary *dic) {
+        return [[dic objectForKey:UNIQUE_ACCOUNT_ID] isEqualToString:uniqueAccountId];
+    }
+                                                       error:error];
+}
+
++ (NSString *)intuneMamResource:(NSURL *)authority error:(ADAuthenticationError *__autoreleasing *) error
```
Comment (jasoncoolmax) [inline ADAL/src/ADEnrollmentGateway.m]: same code style nit as above, also remove space after the right parenthesis :)

File: ADAL/tests/unit/ios/ADEnrollmentGatewayTests.m [hunk 18]
```diff
@@ -0,0 +1,373 @@
...
+                                                                      "]\n"
+                                                                  "}"]];
+}
+
+- (void)tearDown
+{
+    [super tearDown];
+}
+
+- (void) testenrollmentIdForUserId_whenJSONIsCorrect_shouldReturnEnrollmentID
```
Comment (jasoncoolmax) [inline ADAL/tests/unit/ios/ADEnrollmentGatewayTests.m]: nit: remove space after the right parenthesis, same for all other unit tests :)

File: ADAL/src/ADEnrollmentGateway.h [hunk 19]
```diff
@@ -0,0 +1,51 @@
...
+
+/*! Returns all known Intune MAM Enrollments IDs as JSON*/
++ (NSString *)allEnrollmentIdsJSON;
+
+/*! Returns the first available enrollmentID if one is available */
++ (NSString *)enrollmentIdIfAvailable:(ADAuthenticationError * __autoreleasing *)error;
+
+/*! Tries to find an enrollmentID for a homeUserID/uniqueAccoutnID first, then checks userID,
+    then returns any enrollmentID available */
++ (NSString*)enrollmentIDForUniqueAccountID:(NSString*) homeUserID userID:(NSString*) userID error:(ADAuthenticationError * __autoreleasing *)error;
```
Comment (oldalton) [inline ADAL/src/ADEnrollmentGateway.h]: nit: please call it consistently "homeAccountId" both in method name and argument

File: ADAL/src/ADEnrollmentGateway.h [hunk 20]
```diff
@@ -0,0 +1,51 @@
...
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+@interface ADEnrollmentGateway : NSObject
+
++ (NSString *)enrollmentIdForUserId:(NSString *)userId error:(ADAuthenticationError * __autoreleasing *)error;
++ (NSString *)enrollmentIdForUserObjectId:(NSString *)userObjectId tenantId:(NSString *)tenantId error:(ADAuthenticationError * __autoreleasing *)error;
++ (NSString *)enrollmentIdForUniqueAccountId:(NSString *)uniqueAccountId error:(ADAuthenticationError * __autoreleasing *)error;
```
Comment (oldalton) [inline ADAL/src/ADEnrollmentGateway.h]: nit: please rename uniqueAccountId to homeAccountId

File: ADAL/src/ADEnrollmentGateway.m [hunk 21]
```diff
@@ -0,0 +1,257 @@
...
+#define AD_INTUNE_ENROLLMENT_ID_KEY (AD_INTUNE_ENROLLMENT_ID AD_INTUNE_ENROLLMENT_ID_VERSION)
+
+// Keys for Intune Resource
+#define AD_INTUNE_RESOURCE_ID @"intune_mam_resource_V"
+#define AD_INTUNE_RESOURCE_ID_VERSION @"1"
+#define AD_INTUNE_RESOURCE_ID_KEY (AD_INTUNE_RESOURCE_ID AD_INTUNE_RESOURCE_ID_VERSION)
+
+NSString *const ENROLLMENT_ID_ARRAY = @"enrollment_ids";
+
+NSString *const TID = @"tid";
```
Comment (oldalton) [inline ADAL/src/ADEnrollmentGateway.m]: nit: please prefix those constants

File: ADAL/src/ADEnrollmentGateway.m [hunk 22]
```diff
@@ -0,0 +1,257 @@
...
+NSString *const OID = @"oid";
+NSString *const UNIQUE_ACCOUNT_ID = @"unique_account_id";
```
Comment (oldalton) [inline ADAL/src/ADEnrollmentGateway.m]: let's call it consistently homeAccountId. Sorry for confusion :)

File: ADAL/tests/unit/ios/ADEnrollmentGatewayTests.m [hunk 23]
```diff
@@ -0,0 +1,390 @@
...
+@interface ADEnrollmentGatewayTests : ADTestCase
+@end
+
+@implementation ADEnrollmentGatewayTests
+
+- (void)setUp
+{
+    [super setUp];
+
+    [ADEnrollmentGateway setEnrollmentIdsWithJsonBlob:[ADEnrollmentGateway getTestEnrollmentIDJSON]];
```
Comment (oldalton) [inline ADAL/tests/unit/ios/ADEnrollmentGatewayTests.m]: nit: clear ADEnrollmentGateway in tearDown
